Updated JSON Definition File Tree

jafeltra opened the pull request:
This PR updates the file tree that is rendered on the RHS editor.

A few features of the file tree:
- If the JSON entered for that definition is valid, a check mark is displayed next to its name
- If the JSON is invalid, an X is displayed next to it and the text is colored red and bolded (I don't actually think the bolding is necessary since there is a non-color based visual with the X and I didn't mind it, so I left it for others' opinions)
- Choosing a different definition switches the definition loaded in the editor
- Changing the `id` property in the JSON updates the text in the file tree
- Changing the `resourceType` will re-sort the definition. If it is anything besides "StructureDefinition", "ValueSet", or "CodeSystem", it is sorted under "Instances".

A few implementation/code notes:
- I updated the type of `fhirDefinitions` in the `JSONOutput` component to be a string so that any JSON value is stored, not just valid JSON, letting you keep invalid JSON in one definition while viewing another
- The grey background color of selected definitions doesn't fill the name if the name is long enough to require a horizontal scroll. I couldn't figure out how to do this, so if anyone has any ideas, let me know.
- There isn't a test for typing in invalid JSON and testing that the `listItemError` class is added because I can't figure out how to type directly in the editor within the test. Again, if anyone has any ideas, I'm all for them.
- I did set a `style={{ height: '75vh' }}`, which I think will need to change with the changes to the collapsable console. I'm not sure if now is the right time to rebase the `gofsh` branch (and then this one) to have those changes properly included.

ngfreiter reviewed (commented):
Overall this looks good! I have a few comments below.

I'm no designer, but for me I don't think the check mark is really necessary. I don't think we need to signal when the JSON is valid, and can get away with just signaling when it is not valid (with the red x and so on).

I've noticed that the `Loading` text for GoFSH is red, is that intentional:
![image](https://user-images.githubusercontent.com/52747882/111500785-be81a900-871a-11eb-8725-3c0c95a742e5.png)

I've found an issue where if I press the "Run SUSHI" and then "Run GoFSH" buttons in quick succession, I get the following crash:
![image](https://user-images.githubusercontent.com/52747882/111502815-9430eb00-871c-11eb-9de6-731f4c4ea70d.png)

I don't think we have to do anything smart in this case I would be fine with the "Run GoFSH" just not working, or with the "Run GoFSH" running after the "Run SUSHI" has finished. Whatever you think works more smoothly to be honest. I just think we want to avoid the crash, since it's a plausible thing to happen.

jafeltra commented:
@ngfreiter I think all three of your comments should be resolved by the last two commits (e3d95d7 and 379acfa). I got rid of the check mark icon and replaced it with some blank space because I thought it looked weird to have the invalid definition names get shifted to the right to add the X icon. I also updated both editors to display "Loading..." in their placeholder text. I think this is probably how it should have been the whole time. It also let me consistently return either a string (for FSH) or an array of strings (for GoFSH), which also eliminated the crash. Let me know what you think!

jafeltra commented:
Actually.. the placeholder updates might not be perfect. It doesn't look like the text ever gets cleared out to show the "Loading..." placeholder text if there is already text in the editor. I might need to look at that tomorrow morning.

jafeltra commented:
Julia - I regret to inform you, but I have found a bug on your branch. To reproduce it, remove any `id` field from a definition. Disaster! 

![image](https://user-images.githubusercontent.com/30803904/111666814-75008f00-87ea-11eb-80b8-605a17cedc50.png)

cmoesel commented:
> Julia - I regret to inform you, but I have found a bug on your branch. To reproduce it, remove any `id` field from a definition. Disaster!
> 
> ![image](https://user-images.githubusercontent.com/30803904/111666814-75008f00-87ea-11eb-80b8-605a17cedc50.png)

Oh man.  Julia is going to hate you now for finding that.

jafeltra commented:
Never fear @jafeltra and @cmoesel, I've discovered a fix! d009095 should do the trick! Thanks for uncovering this issue!

ngfreiter reviewed (approved):
This looks good to me, especially now that Julia fixed that bug that Julia found.

julianxcarter reviewed (approved):
This is a really cool feature, and it looks great and works well for me. I'm approving, awesome work! @jafeltra
